### ISSUE TITLE
How to include a (C#) internal class?

### ISSUE THREAD

**Issue by dougclutter** [dotnet/docfx#2561]
**DocFX Version Used**: 2.33.0

**Template used**: default

**Steps to Reproduce**:

1. I have an internal class that I would like to include in the API docs
2. Added "include" to filterConfig.yml  (details below)

**Expected Behavior**:
The filterConfig.yml is referenced in the docfx.json file.  We know this is working because we are excluding some classes.

Added the following to filterConfig.yml:  
```
- include:
    uidRegex: ^My\.Internal\.ClassName$
    type: Class
```
Expected the class My.Internal.ClassName to be included in the API docs.

**Actual Behavior**:
My.Internal.ClassName is not included in the API docs.

**Comment by vwxyzh** [comment]
@dougclutter removing internal classes/members is hard code logic in [default filter](https://github.com/dotnet/docfx/blob/dev/src/Microsoft.DocAsCode.Metadata.ManagedReference.Roslyn/Visitors/DefaultFilterVisitor.cs#L114).

**Comment by vicancy** [comment]
Hi @dougclutter could you help us to understand your scenario? why you'd like internal classes included? And why do customer want to see internal classes?

**Comment by Still34** [comment]
While I think internal should rarely be exposed into a public API documentation, sometimes if the XML docs refers to something that's internal, that should probably be addressed. Either by showing the docs that's not shown publicly in the API list or transforming the type into a normal string.

**Comment by dougclutter** [comment]
Hi @vicancy one of the things that we've decided to do is make our Entity Framework DbContext classes internal.  We do this because we don't want programmers working on the UI or API to be able to create arbitrary queries against the database.  However, we're constantly creating business objects that use the DbContext, so we'd like to include it in our docs.

Unlike many of your users, we are generating docs for our internal development team rather than the general public.  Hope this helps!

**Comment by dougclutter** [comment]
@vwxyzh Yes.  I'd like to be able to:  
1. Include specific internal classes that I specify in the filters
2. Include specific private properties/methods that I specify in the filters

**Comment by dougclutter** [comment]
@vicancy Can this be considered a `feature-request` now?

**Comment by murt** [comment]
> could you help us to understand your scenario?

I can also add a use case to this that will become more prevalent in the future as more game engines integrate the CLI and languages beyond C# become supported.

Unity3D provide a means of connecting editor components to publicly accessible members, however they also support private members via the `<SerializeField>` attribute.

This works okay for C# and in all honesty `public` should probably be used for these values (and thus not be an issue with docfx) however F# works a bit differently - we are *forced* to use `let mutable` which is "private"

The following are equivalent:

```csharp
class Foo : MonoBehaviour {
     [<SerializeField>]
     private float bar = 42.0f
}
```

```fsharp
type Foo () =
    inherit MonoBehaviour()

   [<SerializeField>]
   let mutable bar = 42.0f
```



It would be great if I could filter to include private members or even filter for particular attributes (as I am only likely to want to document members with `<SerializeField>` anyways)

**Comment by jchristin** [comment]
FYI `disableDefaultFilter` is already working in v2.37.2 if you add it to docfx.json. A bug is preventing it to work through metadata command line options (see #3147).

**Comment by herohua** [comment]
Thank you @jchristin to help us fix #3147. The fix will be released in v2.37.3.

**Comment by jchristin** [comment]
My pleasure.

**Comment by herohua** [comment]
#3147 released in [v2.38](https://github.com/dotnet/docfx/releases/tag/v2.38).

**Comment by dougclutter** [comment]
@vwxyzh - I tried adding _disableDefaultFilter: true to the docfx.json in the build.globalMetadata section, but it doesn't appear to do anything.  Am I setting it in the right place?

**Comment by dougclutter** [comment]
@vwxyzh - I'm also not sure how the DisableDefaultFilter gets me where I want to be.  It seems that disabling the default filter would include *all* the internal classes; that's definitely *not* what I want.  I want to be able to select specific internal classes.

**Comment by superyyrrzz** [comment]
@dougclutter You should set it in metadata config, e.g.
```json
{
  "metadata": [
    {
      "src": [
        {
          "files": [ "**.csproj" ],
          "exclude": [ "**/bin/**", "**/obj/**" ],
          "src": "../src"
        }
      ],
      "dest": "obj/temp/docfxapi",
      "disableDefaultFilter": true
    }
  ],
  "build": ...
}
```
and [API filter](https://dotnet.github.io/docfx/tutorial/howto_filter_out_unwanted_apis_attributes.html) should be the feature you are looking for.

**Comment by crcrewso** [comment]
One reason I see for building documentation including internal and private classes/methods is for documentation. 

I'm required to produce static documentation aligning to versions, at various points, for future codebase maintenance. I would very much like to use the Docfx to produce this document without having to make everything public.

**Comment by JeremyCaney** [comment]
Which rule in the [Default Filter Configuration](https://dotnet.github.io/docfx/tutorial/howto_filter_out_unwanted_apis_attributes.html#3-default-filter-configuration) is responsible for filtering out `internal` classes? I assume it’s an attribute that gets assigned to `internal` types and members during compilation? I ask because, presumably, this would allow finer control over without disabling _all_ filter rules.

**Comment by jldubz** [comment]
For reference in case it helps anyone else 🧙‍♂️

I had a need to include the documentation for a specific set of internal classes because the documentation on their members was being inherited in other places in the code. I want to keep these internal because they should not be exposed in the resulting library. Without the internal classes included, the other documentation does not correctly inherit. I solved this by:
- Copying the [default filter configuration](https://dotnet.github.io/docfx/tutorial/howto_filter_out_unwanted_apis_attributes.html#3-default-filter-configuration) to a new file in my docfx project. I called it `filterConfig.yml`
- Adding include statements for each class and namespace I needed in addition to what was already being processed
- Disabling the default filter configuration
- Pointing to my new filter configuration
